### PR TITLE
Decouple log reader from diagnostics

### PR DESCRIPTION
**Pull request by GabeDeBacker**
This PR addresses issue #238. Currently the log-reader was not only tied to the diagnostic collection, but also tied to the upgrade SARIF prompt as well.

This change decouples the log-reader from both of those and allows the UI (both the diagnostic collection and the upgrade prompt) to be controlled by callers of the log-reader.

There is still remaining UI decoupling to do (mainly the progress UI that the log-reader drives). Issue #240 has been opened for that.

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -62,3 +72,133 @@ export function deactivate(): void {
...
+    }
+}
+
+/**
+ * When a sarif document closes we need to clear all of the list of issues and reread the open sarif docs
+ * Can't selectivly remove issues becuase the issues don't have a link back to the sarif file it came from
+ * @param doc document that was closed
+ */
+function onDocumentClosed(doc: vscode.TextDocument, diagnosticCollection: SVDiagnosticCollection): void {
+    if (Utilities.isSarifFile(doc)) {
```
Comment (GabeDeBacker) [inline src/extension.ts:201]: >doc [](start = 30, length = 3)

The utilities code shouldn't operate on VSCode UI concepts.

File: src/fileConverter.ts [hunk 2]
```diff
@@ -77,77 +114,89 @@ export class FileConverter {
...
-            // as it open the converted document when it is done whic starts
-            // the read SARIF cycle again.
-            // tslint:disable-next-line: no-floating-promises
-            FileConverter.upgradeSarif(doc, parsedVer, parsedSchemaVer);
-        } else {
-                await vscode.window.showErrorMessage(localize(
-                    'converterTool.UpgraderErrorMessage',
-                    "Sarif version '{0}'(schema '{1}') is not yet supported by the Viewer. Make sure you have the latest extension version and check https://github.com/Microsoft/sarif-vscode-extension for future support.",
-                    version, schema));
+        if (!sarifLog.$schema.startsWith('http://json.schemastore.org/sarif-') &&
```
Comment (roywill) [inline src/fileConverter.ts:170]: >http: [](start = 42, length = 5)

The protocol/scheme is case insensitive I do believe.

File: src/logReader.ts [hunk 3]
```diff
@@ -62,101 +88,86 @@ export class LogReader implements Disposable {
...
-        // Keep a flag that indicates whether we need to send the collected issues to
-        // VSCode's problems pane (i.e. synchronize diagnostics to the diagnostics collection)
-        let synchronizeDiagnosticsCollection: boolean = false;
-
-        // Spin through VSCode's documents and read any SARIF files that are opened.
-        for (const sarifTextDocument of workspace.textDocuments.filter((doc) => Utilities.isSarifFile(doc))) {
-            await this.read(sarifTextDocument, { synchronizeDiagnosticsCollection: false });
-            synchronizeDiagnosticsCollection = true;
+    public async read(sarifFile: Uri): Promise<LogReaderResult> {
+        if (!sarifFile.isFile() && Utilities.isSarifFile(sarifFile.fsPath)) {
```
Comment (roywill) [inline src/logReader.ts:95]: Fix this

File: src/svDiagnosticCollection.ts [hunk 4]
```diff
@@ -212,11 +200,29 @@ export class SVDiagnosticCollection implements Disposable {
...
     }
 
+    /**
+     * Adds read results to the collection of diagnostics.
+     * @param readResults Results to add.
+     */
+    public addReadResults(readResults: ParseResults[]): void {
+        for (const readResult of readResults) {
+            this.runInfoCollection.push(readResult.runInfo);
+            for (const resultInfo of readResult.results) {
```
Comment (roywill) [inline src/svDiagnosticCollection.ts:210]: Rename something here.

File: src/fileConverter.ts (hunk 2 above)
Comment (GabeDeBacker) [reply src/fileConverter.ts:170]: You are correct. Fixing this check.

---
In reply to: [408274949](https://github.com/microsoft/sarif-vscode-extension/pull/241#discussion_r408274949) [](ancestors = 408274949)

**Comment by jeffersonking** [comment]
Reviewed in person with @GabeDeBacker.

**Pull request merged**